DataStructures/MinMaxHeap.cs

```diff
@@ -0,0 +1,405 @@
+        }
+
+        /// <summary>
+        /// Removes the minimum value from the heap.
+        /// </summary>
+        /// <returns>true if value is removed; otherwise false.</returns>
+        public bool RemoveMin()
+        {
+            if (Count == 0)
+            {
+                return false;
+            }
+
+            RemoveNode(0);
+            return true;
+        }
+
+        /// <summary>
+        /// Removes the maximum node from the heap and returns its value.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">Thrown if heap is empty.</exception>
+        /// <returns>Value of the removed maximum node.</returns>
+        public T ExtractMax()
+        {
+            if (Count == 0)
+            {
+                throw new InvalidOperationException("Heap is empty");
+            }
+
+            T max = GetMax();
+            RemoveNode(GetMaxNodeIndex());
+            return max;
+        }
+
+        /// <summary>
+        /// Removes the minimum node from the heap and returns its value.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">Thrown if heap is empty.</exception>
+        /// <returns>Value of the removed minimum node.</returns>
+        public T ExtractMin()
+        {
+            if (Count == 0)
+            {
+                throw new InvalidOperationException("Heap is empty");
+            }
+
+            T min = GetMin();
+            RemoveNode(0);
+            return min;
+        }
+
+        /// <summary>
+        /// Gets the maximum value in the heap, as defined by the comparer.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">Thrown if heap is empty.</exception>
+        /// <returns>The maximum value in the heap.</returns>
+        public T GetMax()
+        {
+            if (Count == 0)
+            {
+                throw new InvalidOperationException("Heap is empty");
+            }
+
+            return Heap[GetMaxNodeIndex()];
+        }
+
+        /// <summary>
+        /// Gets the minimum value in the heap, as defined by the comparer.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">Thrown if heap is empty.</exception>
+        /// <returns>The minimum value in the heap.</returns>
+        public T GetMin()
+        {
+            if (Count == 0)
+            {
+                throw new InvalidOperationException("Heap is empty");
+            }
+
+            return Heap[0];
+        }
+
+        /// <summary>
+        /// Finds maximum or minumum value among children and grandchildren of the specified node.
+        /// </summary>
+        /// <param name="index">Index of the node in the Heap array.</param>
+        /// <param name="minimum">If true find minimum value, otherwise - maximum.</param>
+        /// <returns>Index of the maximum or minimum descendant.</returns>
+        private int ExtremumChildOrGrandchild(int index, bool minimum)
```

Comment:
Could you split this method into IndexOfMinChildOrGrandchild and IndexOfMaxChildOrGrandchild? It's a bad smell to have boolean flags as method parameters. Yes, those methods will differ only in one line, but calling code will be much more readable IMO. What do you think?

Reply:
I was not sure how to implement this method. I agree, splitting will make code more readable.